statsmodels/base/_penalties.py

```diff
@@ -73,54 +86,457 @@ def grad(self, params):
+        p = np.atleast_1d(params)
+        p_abs = np.abs(p)
+        res = np.zeros(p_abs.shape)
+
+        mask1 = p_abs < tau
+        mask3 = p_abs >= self.c * tau
+        mask2 = ~mask1 & ~mask3
+        res[mask2] = -1 / (self.c - 1)
+
+        return self.weights * res
+
+
+class SCADSmoothed(SCAD):
+    """
+    The SCAD penalty of Fan and Li, quadratically smoothed around zero.
+
+    This follows Fan and Li 2001 equation (3.7).
+
+    Parameterization follows Boo, Johnson, Li and Tan 2011
+    see docstring of SCAD
+
+    Parameters
+    ----------
+    tau : float
+        slope and threshold for linear segment
+    c : float
+        factor for second threshold
+    c0 : float
+        threshold for quadratically smoothed segment
+    restriction : None or array
+        linear constraints for
+
+
+    Notes
+    -----
+    TODO: Use delegation instead of subclassing, so smoothing can be added to
+    all penalty classes.
+
+    """
+
+    def __init__(self, tau, c=3.7, c0=None, weights=None, restriction=None):
+        if weights is None:
+            self.weights = 1.
+        else:
+            self.weights = weights
+        self.alpha = 1.
+        self.tau = tau
+        self.c = c
+        self.c0 = c0 if c0 is not None else tau * 0.1
+        if self.c0 > tau:
+            raise ValueError('c0 cannot be larger than tau')
+
+        # get coefficients for quadratic approximation
+        c0 = self.c0
+        deriv_c0 = super(SCADSmoothed, self).grad(c0)
+        value_c0 = super(SCADSmoothed, self).func(c0)
+        self.aq1 = value_c0 - 0.5 * deriv_c0 * c0
+        self.aq2 = 0.5 * deriv_c0 / c0
+        self.restriction = restriction
+
+    def func(self, params):
+        # TODO: `and np.size(params) > 1` is hack for llnull, need better solution
+        if self.restriction is not None and np.size(params) > 1:
+            params = self.restriction.dot(params)
+        # need to temporarily override weights for call to super
+        # Note: we have the same problem with `restriction`
+        weights = self.weights
+        self.weights = 1.
+        value = super(SCADSmoothed, self).func(params[None, ...])
+        self.weights = weights
+
+        # shift down so func(0) == 0
+        value -= self.aq1
+        #change the segment corrsponding to quadratic approximation
+        p_abs = np.atleast_1d(np.abs(params))
+        mask = p_abs < self.c0
+        p_abs_masked = p_abs[mask]
+        value[mask] = self.aq2 * p_abs_masked**2
+
+        return (self.weights * value).sum(0)
+
+    def grad(self, params):
+        if self.restriction is not None and np.size(params) > 1:
+            params = self.restriction.dot(params)
+        # need to temporarily override weights for call to super
+        weights = self.weights
+        self.weights = 1.
+        value = super(SCADSmoothed, self).grad(params)
+        self.weights = weights
+
+        #change the segment corrsponding to quadratic approximation
```

Comment:
Last couple nitpicks in this file -- generally looks good -- this comment and the similar one above could get an extra space after the `#`.  The copy below already has that format.

A few of the docstrings have two blank lines between sections instead of one.  Not a huge deal, but if convenient.